libraries/Microsoft.Bot.Configuration/BotConfigurationExtensions.cs

```diff
@@ -0,0 +1,153 @@
+        /// <returns>Tuple with a boolean that indicates if a bot configuration file's Luis keys are present and its Error message.</returns>
+        public static Tuple<bool, string> ValidateLuisKeyExistence(this BotConfiguration botConfiguration)
+        {
+            // Get all the luis types in the service list
+            var luisServices = botConfiguration.Services.OfType(ServiceTypes.Luis);
+
+            // If there isn't any Luis service, returns false and the corresponding error message
+            if (luisServices == null)
+            {
+                return Tuple.Create(false, "There is no LUIS service in the bot configuration file.");
+            }
+
+            // Check if exists any LuisService without its key (SubscriptionKey)
+            var existEmptyLuis = luisServices.Any(x => string.IsNullOrWhiteSpace(((LuisService)x).SubscriptionKey));
+
+            // If there is at least one LuisService without its key, the process will return false
+            if (existEmptyLuis)
+            {
+                return Tuple.Create(false, "There is at least one LuisService without a key.");
+            }
+
+            // Get all the Dispatch (Luis) types in the service list
+            var dispatchServices = botConfiguration.Services.OfType(ServiceTypes.Dispatch);
+
+            // The dispatch service is not always included in the the bot configuration files.
+            // Therefore, it will only be validated if the bot configuration file contains at least one service of this type.
+            if (dispatchServices.Count() != 0)
+            {
+                // Check if exists any dispatchServices without its key (SubscriptionKey)
+                var existEmptyDispatch = dispatchServices.Any(x => string.IsNullOrWhiteSpace(((DispatchService)x).SubscriptionKey));
+
+                string errMsg = existEmptyDispatch ? "There is at least one ServiceDispatch without a key." : string.Empty;
+
+                // If there is at least one dispatch service without its key, the process will return false and the corresponding error message
+                return Tuple.Create(!existEmptyDispatch, errMsg);
+            }
+
+            return Tuple.Create(true, string.Empty);
+        }
+
+        /// <summary>
+        /// Checks if the <see cref="QnAMakerService"/> has its key defined.
+        /// </summary>
+        /// <param name="botConfiguration">BotConfiguration object to extend.</param>
+        /// <returns>Boolean that indicates if a bot configuration file's QnAMaker keys are present.</returns>
+        public static Tuple<bool, string> ValidateQnAKeyExistence(this BotConfiguration botConfiguration)
+        {
+            // Get all the QnA types in the service list
+            var qnaServices = botConfiguration.Services.OfType(ServiceTypes.QnA);
+
+            // If there isn't any qna service, returns false and the corresponding error message
+            if (qnaServices.Count() == 0)
```

Comment:
Again, `.Count()` is `O(N)`... you should really just do your `.Any` check from below right here.